AdminServer/appscale/admin/service_manager.py

```diff
@@ -0,0 +1,409 @@
+          time.time() > server.failure_time + self.FAILED_SERVER_RETENTION):
+        return True
+
+      if server.state == ServerStates.STOPPED:
+        return True
+
+      return False
+
+    self.state = [server for server in self.state if not outdated(server)]
+    for service_type, options in self.assignments.items():
+      yield self._schedule_service(service_type, options)
+
+    for server in self.state:
+      if server.state != ServerStates.RUNNING:
+        continue
+
+      IOLoop.current().spawn_callback(server.ensure_running)
+
+  def _get_open_port(self):
+    """ Selects an available port for a server to use.
+
+    Returns:
+      An integer specifying a port.
+    """
+    assigned_ports = set(service.port for service in self.state)
+    port = self.START_PORT
+    while True:
+      # Skip ports that have been assigned.
+      if port in assigned_ports:
+        port += 1
+        continue
+
+      return port
+
+  @gen.coroutine
+  def _schedule_service(self, service_type, options):
+    """ Schedules servers to fulfill service assignment.
+
+    Args:
+      service_type: A string specifying the service type.
+      options: A dictionary specifying options to use when starting servers.
+    """
+    scheduled = [server for server in self.state
+                 if server.type == service_type and
+                 server.state in self.SCHEDULED_STATES]
+    to_start = options['count'] - len(scheduled)
+    if to_start < 0:
+      stopped = 0
+      for server in reversed(scheduled):
+        if stopped >= abs(to_start):
+          break
+
+        logger.info('Stopping {}'.format(server))
+        IOLoop.current().spawn_callback(server.stop)
+        stopped += 1
+
+      return
+
+    for _ in range(to_start):
+      port = self._get_open_port()
+      server_class = self.SERVICE_MAP[service_type]['server']
+      server = server_class(port, self._http_client, options['verbose'])
+      self.state.append(server)
+      logger.info('Starting {}'.format(server))
+      IOLoop.current().spawn_callback(server.start)
+
+  @gen.coroutine
+  def _update_services(self, assignments):
+    """ Updates service schedules to fulfill assignments.
+
+    Args:
+      assignments: A dictionary specifying service assignments.
+    """
+    self.assignments = assignments
+    for service_type, options in assignments.items():
+      yield self._schedule_service(service_type, options)
+
+  def _update_services_watch(self, encoded_assignments, _):
+    """ Updates service schedules to fulfill assignments.
+
+    Args:
+      encoded_assignments: A JSON-encoded string specifying service
+        assignments.
+    """
+    persistent_update_services = retry_data_watch_coroutine(
+      self._assignments_path, self._update_services)
+    try:
+      assignments = json.loads(encoded_assignments)
+    except (TypeError, ValueError):
+      if encoded_assignments:
+        raise
+
+      assignments = {}
```

Comment:
Isn't it better to check if `encoded_assignments` is empty before loading json instead of `try-except`?
```
assignments = json.loads(encoded_assignments) if encoded_assignments else {}
```

Reply:
Good point. That is more concise and probably more readable too.